lib/deploy/data.ts

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    GitProject,
+    HandlerContext,
+    logger,
+    Project,
+} from "@atomist/automation-client";
+import {
+    GoalInvocation,
+    SdmGoalEvent,
+} from "@atomist/sdm";
+import { readSdmVersion } from "@atomist/sdm-core";
+import * as k8s from "@kubernetes/client-node";
+import * as dockerfileParser from "docker-file-parser";
+import * as stringify from "json-stringify-safe";
+import * as _ from "lodash";
+import { DeepPartial } from "ts-essentials";
+import { defaultNamespace } from "../kubernetes/namespace";
+import { KubernetesApplication } from "../kubernetes/request";
+import {
+    goalEventSlug,
+    KubernetesDeploy,
+    KubernetesDeployRegistration,
+} from "./goal";
+import { loadKubernetesSpec } from "./spec";
+
+/**
+ * JSON propery under the goal event data where the
+ * [[KubernetesDeployment]] goal [[KubernetesApplication]] data are
+ * stored, i.e., the value of `goal.data[sdmPackK8s]`.
+ */
+export const sdmPackK8s = "sdm-pack-k8s";
+
+/**
+ * Generate KubernetesApplication from project and goal invocation.
+ * This function used [[defaultDeploymentData]] to produce initial
+ * [[KubernetesApplication]] data.  It then merges in any object
+ * provided by `goalEvent.data[[[sdmPackk8]]]`, with the latter taking
+ * precedence.  Then, if `registration.applicationData` is truthy, it
+ * calls that function, passing the merged value of the Kubernetes
+ * application, and uses its return value as
+ * `goalEvent.data[[[sdmPackk8]]]`.
+ *
+ * @param k8Deploy Kubernetes deployment goal
+ * @param registration Goal registration/configuration for `k8Deploy`
+ * @param goalInvocation The Kubernetes deployment goal currently triggered.
+ * @return The SdmGoalEvent augmented with [[KubernetesApplication]] in the data
+ */
+export function generateKubernetesGoalEventData(
+    k8Deploy: KubernetesDeploy,
+    registration: KubernetesDeployRegistration,
+    goalInvocation: GoalInvocation,
+): Promise<SdmGoalEvent> {
+    return k8Deploy.sdm.configuration.sdm.projectLoader.doWithProject({ ...goalInvocation, readOnly: true }, async p => {
```

Comment:
Can this use the `doWithProject` function directly?

Reply:
The action passed to [`doWithProject`](https://github.com/atomist/sdm/blob/222bbf43b704ef4e76daf2f8e422217b67c2ed80/lib/api-helper/project/withProject.ts#L79) returns a `Promise<ExecuteGoalResult>` whereas here we are returning a `Promise<SdmGoalEvent>`?

Reply:
Right, that won’t work.

Reply:
Maybe we should make `doWithProject` return `Promise<T>` so it can be used anywhere?